consensus/view_change.go

```diff
@@ -214,228 +188,73 @@ func (consensus *Consensus) onViewChange(msg *msg_pb.Message) {
+			Msg("Verify View Change Message Error")
+		return
 	}
 
-	// m2 type message
-	if !hasBlock {
-		_, ok := consensus.nilSigs[recvMsg.ViewID][senderKey.Bytes.Hex()]
-		if ok {
-			consensus.getLogger().Debug().
-				Str("validatorPubKey", senderKey.Bytes.Hex()).
-				Msg("[onViewChange] Already Received M2 message from validator")
-			return
-		}
-
-		if !recvMsg.ViewchangeSig.VerifyHash(senderKey.Object, NIL) {
-			consensus.getLogger().Warn().Msg("[onViewChange] Failed To Verify Signature For M2 Type Viewchange Message")
-			return
-		}
-
-		consensus.getLogger().Info().
-			Str("validatorPubKey", senderKey.Bytes.Hex()).
-			Msg("[onViewChange] Add M2 (NIL) type message")
-		consensus.nilSigs[recvMsg.ViewID][senderKey.Bytes.Hex()] = recvMsg.ViewchangeSig
-		consensus.nilBitmap[recvMsg.ViewID].SetKey(recvMsg.SenderPubkey.Bytes, true) // Set the bitmap indicating that this validator signed.
-	} else { // m1 type message
-		if consensus.BlockVerifier(preparedBlock); err != nil {
-			consensus.getLogger().Error().Err(err).Msg("[onViewChange] Prepared block verification failed")
-			return
-		}
-		_, ok := consensus.bhpSigs[recvMsg.ViewID][senderKey.Bytes.Hex()]
-		if ok {
-			consensus.getLogger().Debug().
-				Str("validatorPubKey", senderKey.Bytes.Hex()).
-				Msg("[onViewChange] Already Received M1 Message From the Validator")
-			return
-		}
-		if !recvMsg.ViewchangeSig.VerifyHash(recvMsg.SenderPubkey.Object, recvMsg.Payload) {
-			consensus.getLogger().Warn().Msg("[onViewChange] Failed to Verify Signature for M1 Type Viewchange Message")
-			return
-		}
-
-		// first time receive m1 type message, need verify validity of prepared message
-		if len(consensus.m1Payload) == 0 || !bytes.Equal(consensus.m1Payload, recvMsg.Payload) {
-			if len(recvMsg.Payload) <= 32 {
-				consensus.getLogger().Warn().
-					Int("len", len(recvMsg.Payload)).
-					Msg("[onViewChange] M1 RecvMsg Payload Not Enough Length")
-				return
-			}
-			blockHash := recvMsg.Payload[:32]
-			aggSig, mask, err := consensus.ReadSignatureBitmapPayload(recvMsg.Payload, 32)
-			if err != nil {
-				consensus.getLogger().Error().Err(err).Msg("[onViewChange] M1 RecvMsg Payload Read Error")
-				return
-			}
-
-			if !consensus.Decider.IsQuorumAchievedByMask(mask) {
-				consensus.getLogger().Warn().
-					Msgf("[onViewChange] Quorum Not achieved")
-				return
-			}
-
-			// Verify the multi-sig for prepare phase
-			if !aggSig.VerifyHash(mask.AggregatePublic, blockHash[:]) {
-				consensus.getLogger().Warn().
-					Hex("blockHash", blockHash).
-					Msg("[onViewChange] failed to verify multi signature for m1 prepared payload")
-				return
+	if msgType == M1 {
+		preparedMsgs := consensus.FBFTLog.GetMessagesByTypeSeq(
+			msg_pb.MessageType_PREPARED, recvMsg.BlockNum,
+		)
+		// search for local prepared msg and add one if not found
+		preparedMsg := consensus.FBFTLog.FindMessageByViewID(preparedMsgs, recvMsg.ViewID)
+		if preparedMsg == nil {
```

Comment:
This should be the original condition to add new prepared block:
`if len(consensus.m1Payload) == 0 || !bytes.Equal(consensus.m1Payload, recvMsg.Payload) {`

new logic is different now.

Reply:
Please double check the logic about adding a m1/prepared block to local fbft log.

Reply:
I've commented on the old PR.  The original code, the m1payload is always set to payload found on the local prepared message at first. And it only sets to payload from recvMsg when it is empty which means it can't find the local prepared message.  So the new logic is the same, only add a local block when it can't find the prepared message from the local fbft log.

The validator should always trust the local message than the received message. that's my understanding.  We can further discuss it.

Reply:
ok, got it. thanks.